packages/dashboard/src/hooks/useWidgetDataTypeSet.ts

```diff
@@ -0,0 +1,21 @@
+import type { SiteWiseAssetQuery } from '@iot-app-kit/source-iotsitewise';
+import { useAssetDescriptionMapAsync } from './useAssetDescriptionMapAsync';
+import { isDefined } from '~/util/isDefined';
+
+export const useWidgetDataTypeSet = (siteWiseAssetQuery: SiteWiseAssetQuery | undefined): Set<string> => {
+  const describedAssetsMap = useAssetDescriptionMapAsync(siteWiseAssetQuery);
+
+  const dataTypes = siteWiseAssetQuery?.assets
+    .map(({ assetId, properties }) =>
+      properties.map(
+        ({ propertyId }) =>
+          describedAssetsMap[assetId]?.assetProperties
+            ?.filter((prop) => prop.id === propertyId)
+            .map((prop) => prop.dataType) || []
+      )
+    )
+    .flat(2)
+    .filter(isDefined);
```

Comment:
This is hard to follow due to the high density of logic. Consider improving the readability.